### DISCUSSION TITLE
Right way to communicate bad request details

### DISCUSSION THREAD

**Discussion by mmkal** [trpc/trpc#1696] [Ideas]
Hey! I'm curious if trpc has any opinions on, or would like to have opinions on, how to communicate what a client did wrong in a "bad request" scenario. For example, if a client sends an object like `POST /createUser -d "{ username: 'abc', password: 'abc123' }"`, it might be nice to have a "blessed" way to send back some extra information from the server about why the request gets rejected. I could imagine it looking something like:

```ts
throw new trpc.TRPCError({
  code: 'BAD_REQUEST',
  message: 'Password is invalid',
  detailsThatAreDefinitelySafeToSendToTheClient: {
    problems: [
      { ruleId: 'length', expected: 8, actual: 6 },
      { ruleId: 'username', expected: 'Not to contain username', actual: 'Contains username' },
    ]
  },
})
```

Where `detailsThatAreDefinitelySafeToSendToTheClient` is a deliberately bad name, but basically can give the client some hints about what error to show to the user/send to error reporting services. It could either be typed as `unknown` or `any` or something, or it could be part of some kind of type config for the router (I know typing errors has been discussed in a few other issues, but it's a nice-to-have for something like this, IMO).

**Comment by jack-douglas** [comment]
You can pass `cause` as an option. https://github.com/trpc/trpc/blob/b85b465a21db63a9adb63e9430781c4f2e1dcfb0/packages/server/src/TRPCError.ts#L19

**Reply by KATT** [reply]
Yeah, I'd use cause for this or just throw my own errors (that end up in cause), then do custom error formatting 

https://trpc.io/docs/error-formatting
